### ISSUE TITLE
Lifetime of key in serialize_struct_elt

### ISSUE THREAD

**Issue by dbrgn** [serde-rs/serde#708]
When implementing `Serialize` for a custom struct, the rough workflow seems to be like this:

```rust
let mut state = serializer.serialize_struct("MyStruct", ...)`
serializer.serialize_struct_elt(&mut state, "key", &val)?;
serializer.serialize_struct_elt(&mut state, "key", &val)?;
...
serializer.serialize_struct_end(state)
```

The `serialize_struct_elt` method seems to require a static lifetime for the key though:

```rust
/// Serializes an element of a struct.
///
/// By default, struct elements are serialized as a map element
/// with the field name as the key.
#[inline]
fn serialize_struct_elt<V>(&mut self, key: &'static str, value: V)
    -> Result<(), Self::Error> where V: Serialize {
    self.serialize_map_elt(key, value)
}
```
(Source: http://tyoverby.com/wire/src/serde/ser/mod.rs.html#303-313)

I'm trying to serialize a map to a struct, where each key in the map (a String) results in a field in the serialized struct.

Therefore I'd like to do something like this:

```rust
for (name, value) in self.map.iter() {
    serializer.serialize_struct_elt(&mut state, name, value);
}
```

...but that is rejected because of that 'static lifetime requirement.

I can't call `serialize_map_elt` directly either though, because it is not implemented on type `S: Serializer`.

How do I dynamically insert struct keys?

**Comment by dtolnay** [comment]
Serde's notion of a struct is something where you know the fields at compile time. Any other key/value pairing is probably a map. Can you clarify why you need to use serialize_struct rather than serialize_map in your case?

The `wire` crate you linked to depends on an old version of Bincode which uses serde 0.7. The map serialization API was changed a bit in 0.8 to support streaming use cases better. Here is an example of using the new API: https://serde.rs/serialize-map.html

**Comment by dbrgn** [comment]
> Serde's notion of a struct is something where you know the fields at compile time. Any other key/value pairing is probably a map. Can you clarify why you need to use serialize_struct rather than serialize_map in your case?

I think in this case I actually do need a map, thanks. Unfortunately the documentation on this is a bit lacking, see #658 and #639. My question in this case would also be what the difference between the two methods is.

> The wire crate you linked to depends on an old version of Bincode which uses serde 0.7. 

Ah, I actually found that code snippet via Google and didn't realize I was looking at the docs for another crate :)

> The map serialization API was changed a bit in 0.8 to support streaming use cases better. Here is an example of using the new API: https://serde.rs/serialize-map.html

Thanks, seems to work in my codebase.

I think this can be closed in favor of #639 and #658.
